src/bp/orchestrator/master.ts

```diff
@@ -0,0 +1,171 @@
+import sdk from 'botpress/sdk'
+import cluster, { Worker } from 'cluster'
+import _ from 'lodash'
+import nanoid from 'nanoid'
+import nanoidGenerate from 'nanoid/generate'
+import yn from 'yn'
+
+import { registerActionServerMainHandler } from './action-server'
+import { registerNluServerMainHandler, killNluProcess } from './nlu-server'
+import { registerStudioMainHandler, killStudioProcess } from './studio-client'
+import { spawnWebWorker, onWebWorkerExit } from './web-worker'
+
+export enum WorkerType {
+  WEB = 'WEB_WORKER',
+  LOCAL_ACTION_SERVER = 'LOCAL_ACTION_SERVER',
+  LOCAL_STAN_SERVER = 'LOCAL_STAN_SERVER',
+  TRAINING = 'TRAINING'
+}
+
+export enum MessageType {
+  StartStudio = 'START_STUDIO',
+  StartActionServer = 'START_ACTION_SERVER',
+  StartNluServer = 'START_STAN_SERVER',
+  RegisterProcess = 'REGISTER_PROCESS',
+  BroadcastProcess = 'BROADCAST_PROCESS',
+  RestartServer = 'RESTART_SERVER'
+}
+
+export type ProcType = 'web' | 'nlu' | 'action-server' | 'studio'
+
+interface SubProcesses {
+  [type: string]: {
+    port: number
+    rebootCount: number
+    workerId?: number
+  }
+}
+
+interface ProcessDetails {
+  processType: ProcType
+  signal?: string | null
+  code?: number | null
+  workerId?: number
+  logger: sdk.Logger
+  exitedAfterDisconnect?: boolean
+  /** If that subprocess fails after the max number of reboots, kill the main process */
+  killOnFail?: boolean
+  /** Optional method to restart the process if it is still restartable */
+  restartMethod?: Function
+}
+
+const debug = DEBUG('cluster')
+const msgHandlers: { [messageType: string]: (message: any, worker: cluster.Worker) => void } = {}
+const maxServerReboots = process.core_env.BP_MAX_SERVER_REBOOT || 2
+
+/**
+ * The master process handles training and rebooting the server.
+ * The worker process runs the actual server
+ *
+ * Exit code 0 or undefined: Success (kill worker & master)
+ * Exit code 1: Error (will try to respawn workers)
+ */
+export const registerMsgHandler = (messageType: string, handler: (message: any, worker: cluster.Worker) => void) => {
+  msgHandlers[messageType] = handler
+}
+
+export const processes: SubProcesses = {}
+
+export const registerProcess = (processType: ProcType, port: number, workerId?: number) => {
+  if (processes[processType]) {
+    processes[processType] = { port, rebootCount: processes[processType].rebootCount + 1 }
```

Comment:
It's to standardize the "reboot" process, since we have the nlu, web worker and studio which can crash & restart, but we don't want to do this indefinitely